### PR TITLE
add timeout and middlewire

### PR DESCRIPTION
**Pull request by wangdayong228**
add middlewire for hook call rpc and batch call rpc
add request timeout for call rpc and batch call rpc

---
This change is [Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)

### CODE REVIEW

**Comment by Thegaram** [comment]
fyi I believe this is called "middleware"

**Comment by wangdayong228** [comment]
> fyi I believe this is called "middleware"

yes, thank you, fixed

**Review by Pana — CHANGES_REQUESTED** [review]
> Reviewed 8 of 13 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 4 unresolved discussions (waiting on @boqiu and @wangdayong228)

*[a discussion](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaS8bbVAL5cWE5ZqgpI:-MaS8bbVAL5cWE5ZqgpJ:b-g6d708) (no related file):*
It's better to add documentation about how to use middleware.
Also, add some logs in the changelog file

---
___
*[api.md, line 239 at r2](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaS3O9mCDJMvzcL06Yo:-MaS3O9n3T4r3EUwUaVJ:b-d1ej3z) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d33ea7d2ed9ab7b5e42f8a10de3f39bbc8296bc0/api.md#L239)):*
> ```Markdown
> Note that batch calls may not be executed atomically on the server side.
> 
> You could use UseBatchCallRpcMiddleWire to add middlewire for hooking
> ```

There is also a typo error here
___
*[client.go, line 49 at r2](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaSFFFD6lztEKHhic7-:-MaSFFFD6lztEKHhic70:b-revj5u) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d33ea7d2ed9ab7b5e42f8a10de3f39bbc8296bc0/client.go#L49)):*
> ```Go
> 	RequestTimeout time.Duration
> 	// log
> 	// CallRpcLog      func(method string, args []interface{}, result interface{}, resultError error, duration time.Duration)
> ```

If these codes won't use anymore, they should be removed 
___
*[go.mod, line 12 at r2](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaS41JG5EkT-BNEfst3:-MaS41JG5EkT-BNEfst4:bq60uan) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d33ea7d2ed9ab7b5e42f8a10de3f39bbc8296bc0/go.mod#L12)):*
> ```XML
> 	github.com/davecgh/go-spew v1.1.1
> 	github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea
> 	github.com/docker/docker v1.4.2-0.20180625184442-8e610b2b55bf
> ```

why is docker package required?

**Review by boqiu — CHANGES_REQUESTED** [review]
> Reviewed 10 of 10 files at r3.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 17 unresolved discussions (waiting on @wangdayong228)

___
*[client.go, line 110 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWJNZz0oj31OA83YtT:-MaWJNZz0oj31OA83YtU:b-ytc47u) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L110)):*
> ```Go
> 	// if client.option.BatchCallRPCLog == nil {
> 	// 	client.option.BatchCallRPCLog = func(b []rpc.BatchElem, err error, duration time.Duration) {}
> 	// }
> ```

Please remove unused code instead of comment.
___
*[client.go, line 160 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWK7jwAW8GWKeKFxhm:-MaWK7jwAW8GWKeKFxhn:bfr5hxs) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L160)):*
> ```Go
> }
> 
> func (client *Client) callRpc(result interface{}, method string, args ...interface{}) error {
> ```

Not used?
___
*[client.go, line 170 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWKIKp3uacmjXstLI7:-MaWKIKp3uacmjXstLI8:b6d38pk) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L170)):*
> ```
> UseCallRpcMiddleware
> ```

Generally, `SetXxx` is used to set any field.
___
*[client.go, line 187 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWKe3J9iuJ_WHHQdQm:-MaWKe3J9iuJ_WHHQdQn:b-hdyhn4) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L187)):*
> ```Go
> }
> 
> func (client *Client) batchCallRPC(b []rpc.BatchElem) error {
> ```

not used?
___
*[client.go, line 1126 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWKtRd9rCroRePGySr:-MaWKtRd9rCroRePGySs:b-p2tv9m) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L1126)):*
> ```Go
> func (client *Client) genContext() context.Context {
> 	var ctx context.Context
> 	ctx = context.Background()
> ```

Why not declare in one line?
___
*[client.go, line 1128 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWL4OdBdDIpdoTez_f:-MaWL4OdBdDIpdoTez_g:bu29vyp) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L1128)):*
> ```Go
> 	ctx = context.Background()
> 	if client.option.RequestTimeout > 0 {
> 		ctx, _ = context.WithTimeout(context.Background(), client.option.RequestTimeout)
> ```

Why the 2nd parameter not used? If so, there will be resource leak.
___
*[go.mod, line 18 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWLOdO1DwcuGL7lZUX:-MaWLOdO1DwcuGL7lZUY:br4baap) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/go.mod#L18)):*
> ```XML
> 	github.com/stretchr/testify v1.5.1 // indirect
> 	github.com/valyala/fasthttp v1.13.1
> 	google.golang.org/protobuf v1.25.0 // indirect
> ```

Please remove all useless `indirect` dependencies.
___
*[interface.go, line 144 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWLZqB8GxJ_gx6hQiE:-MaWLZqB8GxJ_gx6hQiF:beik9l7) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/interface.go#L144)):*
> ```Go
> type RpcRequester interface {
> 	Call(resultPtr interface{}, method string, args ...interface{}) error
> 	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
> ```

It seems the correct API should be `CallTimeout(d time.Duration, ...)`
___
*[rpc/http.go, line 113 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMltWBu6Bl0B8PUen:-MaWMltWBu6Bl0B8PUeo:bpmdwau) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/9a3e8d20b453045fb94deee3350fe6c912ee874e/rpc/http.go#L113)):*
> ```Go
> 	// }
> 
> 	// req := fasthttp.Request{}
> ```

Why `fasthttp` removed?
___
*[rpc/http.go, line 143 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMvw-EdNzvMCFLFIh:-MaWMvw-EdNzvMCFLFIi:bwc21z0) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/9a3e8d20b453045fb94deee3350fe6c912ee874e/rpc/http.go#L143)):*
> ```Go
> 	respBody, err := hc.doRequest(ctx, msg)
> 	// if respBody != nil {
> 	// 	defer respBody.Close()
> ```

`defer` not required?
___
*[middleware/call_rpc.go, line 32 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMAyF7MZxFGQ-7Yh9:-MaWMAyF7MZxFGQ-7YhA:b-p5syhw) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc.go#L32)):*
> ```Go
> 
> 		duration := time.Since(start)
> 		argsJson, _ := json.Marshal(args)
> ```

Why if error occurred?
___
*[middleware/call_rpc.go, line 35 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMIz852eWynBvrdxN:-MaWMIz852eWynBvrdxO:b88ycx) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc.go#L35)):*
> ```Go
> 
> 		if err == nil {
> 			fmt.Printf("%v Method %v, Params %v, Result %v, Use %v\n",
> ```

Do not write any `fmt.Print` in SDK.
___
*[middleware/call_rpc_batch.go, line 32 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMULh2twT3h2CFmHy:-MaWMULh2twT3h2CFmHz:b-c8jlx0) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc_batch.go#L32)):*
> ```
> fmt.Printf
> ```

Again, SDK **must** avoid to output to console. It's up to client to decide how to handle  the failures or add info logs for success.

**Review by boqiu — COMMENTED** [review]
> Reviewed 5 of 13 files at r1, 1 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 17 unresolved discussions (waiting on @wangdayong228)

**Review by wangdayong228 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: 13 of 16 files reviewed, 17 unresolved discussions (waiting on @boqiu and @Pana)

*[a discussion](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaS8bbVAL5cWE5ZqgpI:-MaXPvZG-b7RKPM6QaAm:b-896fix) (no related file):*
<details><summary><i>Previously, Pana (W) wrote…</i></summary><blockquote>

It's better to add documentation about how to use middleware.
Also, add some logs in the changelog file
</blockquote></details>

Done.

---
___
*[api.md, line 239 at r2](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaS3O9mCDJMvzcL06Yo:-MaXEFsh8itBiHQNIm-e:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d33ea7d2ed9ab7b5e42f8a10de3f39bbc8296bc0/api.md#L239)):*
<details><summary><i>Previously, Pana (W) wrote…</i></summary><blockquote>

There is also a typo error here
</blockquote></details>

Done.
___
*[client.go, line 49 at r2](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaSFFFD6lztEKHhic7-:-MaXEHnE06cDPUm1Q3qR:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d33ea7d2ed9ab7b5e42f8a10de3f39bbc8296bc0/client.go#L49)):*
<details><summary><i>Previously, Pana (W) wrote…</i></summary><blockquote>

If these codes won't use anymore, they should be removed 
</blockquote></details>

Done.
___
*[client.go, line 110 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWJNZz0oj31OA83YtT:-MaXEXey2wMIqG_y9VjU:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L110)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Please remove unused code instead of comment.
</blockquote></details>

Done.
___
*[client.go, line 160 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWK7jwAW8GWKeKFxhm:-MaXEaUB6X-l8pJp7vcx:b6sujoi) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L160)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Not used?
</blockquote></details>

It's  the initial middleware function of client.callRpcHandler 
___
*[client.go, line 170 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWKIKp3uacmjXstLI7:-MaXFHOc51Sop_A19fD4:b9un6op) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L170)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

> ```
> UseCallRpcMiddleware
> ```

Generally, `SetXxx` is used to set any field.
</blockquote></details>

Middleware usually use `use`, because it could use more middleware; set means it will be replaced after set again
___
*[client.go, line 187 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWKe3J9iuJ_WHHQdQm:-MaXFUeV2aI9vAxoNQ3w:br2qzj6) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L187)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

not used?
</blockquote></details>

same with callRpc
___
*[client.go, line 1126 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWKtRd9rCroRePGySr:-MaXFfZz3ubX3V7WxIhN:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L1126)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Why not declare in one line?
</blockquote></details>

Done.
___
*[client.go, line 1128 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWL4OdBdDIpdoTez_f:-MaXHSdz4i0aiP6BGWGT:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/client.go#L1128)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Why the 2nd parameter not used? If so, there will be resource leak.
</blockquote></details>

Done.
___
*[go.mod, line 12 at r2](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaS41JG5EkT-BNEfst3:-MaXQ3Bm4u7n4xByPrws:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d33ea7d2ed9ab7b5e42f8a10de3f39bbc8296bc0/go.mod#L12)):*
<details><summary><i>Previously, Pana (W) wrote…</i></summary><blockquote>

why is docker package required?
</blockquote></details>

Done.
___
*[go.mod, line 18 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWLOdO1DwcuGL7lZUX:-MaXJ71i7cJ5gVyo8u-h:b-8e5um0) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/go.mod#L18)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Please remove all useless `indirect` dependencies.
</blockquote></details>

it's used by go-ethereum, they exists because go-ethereum doesn't indicate their go mod version, and I have use `go mod tidy` cleaned it, is it harmful?
___
*[interface.go, line 144 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWLZqB8GxJ_gx6hQiE:-MaXJyLn2Uxm1JzdDaDb:b-r7l5a3) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/interface.go#L144)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

It seems the correct API should be `CallTimeout(d time.Duration, ...)`
</blockquote></details>

CallContext is the go-ethereum rpc Client method name
___
*[rpc/http.go, line 113 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMltWBu6Bl0B8PUen:-MaXO9RZ-6ruzJqjTjSa:b-9q0s9u) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/9a3e8d20b453045fb94deee3350fe6c912ee874e/rpc/http.go#L113)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Why `fasthttp` removed?
</blockquote></details>

just removed commented code
___
*[rpc/http.go, line 143 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMvw-EdNzvMCFLFIh:-MaXOyhj0Utf_Ax5lpmX:b-p60jfg) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/9a3e8d20b453045fb94deee3350fe6c912ee874e/rpc/http.go#L143)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

`defer` not required?
</blockquote></details>

http doesn't have capability to control tcp connection
___
*[middleware/call_rpc.go, line 32 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMAyF7MZxFGQ-7Yh9:-MaXLAbEBbXiyYiwWEMh:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc.go#L32)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Why if error occurred?
</blockquote></details>

Done.
___
*[middleware/call_rpc.go, line 35 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMIz852eWynBvrdxN:-MaXLQygB8Ptd60paLc3:b-v1098b) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc.go#L35)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Do not write any `fmt.Print` in SDK.
</blockquote></details>

It's the default logger middleware, user can customize middleware if need.
___
*[middleware/call_rpc_batch.go, line 32 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMULh2twT3h2CFmHy:-MaXQBXZ35j-BTeF92xh:by9m7e7) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc_batch.go#L32)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

> ```
> fmt.Printf
> ```

Again, SDK **must** avoid to output to console. It's up to client to decide how to handle  the failures or add info logs for success.
</blockquote></details>

 It's the default logger middleware, user can customize middleware if need.

**Review by Pana — DISMISSED** [review]
> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 13 unresolved discussions (waiting on @boqiu)

**Review by boqiu — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 5 unresolved discussions (waiting on @wangdayong228)

___
*[client.go, line 157 at r4](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaXjPztBx9BEr-7KSkc:-MaXjPztBx9BEr-7KSkd:bu56rro) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d44ae9a1685476aac1edeb43e1b3838ff6afed2b/client.go#L157)):*
> ```Go
> 			cancelFunc()
> 		}
> 	}()
> ```

why not execute defer statement only when `cancelfunc != nil`?
___
*[go.mod, line 18 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWLOdO1DwcuGL7lZUX:-MaXkbpg1XF_d11_nsQk:b-e4wdl) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/go.mod#L18)):*
<details><summary><i>Previously, wangdayong228 wrote…</i></summary><blockquote>

it's used by go-ethereum, they exists because go-ethereum doesn't indicate their go mod version, and I have use `go mod tidy` cleaned it, is it harmful?
</blockquote></details>

Basically, **indirect** dependent libs could be removed in **.mod** file.
___
*[middleware/call_rpc.go, line 32 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMAyF7MZxFGQ-7Yh9:-MaXkyn3Dj8IeOvU0LES:bd45fdu) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc.go#L32)):*
<details><summary><i>Previously, wangdayong228 wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Suggest not to ignore the marshal error, instead, please handle the error explicitly.
___
*[middleware/call_rpc.go, line 35 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMIz852eWynBvrdxN:-MaXlFG-8Ihv3nzd8IJd:b-76ylf2) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc.go#L35)):*
<details><summary><i>Previously, wangdayong228 wrote…</i></summary><blockquote>

It's the default logger middleware, user can customize middleware if need.
</blockquote></details>

Generally, the default behavior should not print any error message in stdout. Instead, you could return a stacked error or add any debug level log.

Just remind, no app will print error message to console by default.
___
*[middleware/call_rpc_batch.go, line 32 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMULh2twT3h2CFmHy:-MaXlsXy9FutWw4xNXJ2:b-kfobv4) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc_batch.go#L32)):*
<details><summary><i>Previously, wangdayong228 wrote…</i></summary><blockquote>

 It's the default logger middleware, user can customize middleware if need.
</blockquote></details>

Please see comment above.

**Review by boqiu — COMMENTED** [review]
> Reviewed 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 5 unresolved discussions (waiting on @wangdayong228)

**Review by wangdayong228 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 5 unresolved discussions (waiting on @boqiu and @wangdayong228)

___
*[client.go, line 157 at r4](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaXjPztBx9BEr-7KSkc:-MaXzKrdC1g9jJP6N2FB:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/d44ae9a1685476aac1edeb43e1b3838ff6afed2b/client.go#L157)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

why not execute defer statement only when `cancelfunc != nil`?
</blockquote></details>

Done.
___
*[go.mod, line 18 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWLOdO1DwcuGL7lZUX:-MaY-krx4qdWqFMUr21_:b-fsep0v) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/go.mod#L18)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Basically, **indirect** dependent libs could be removed in **.mod** file.
</blockquote></details>

Do not found document explain the indirect dependent should removed from .mod file, could you suggest related document?

**Review by boqiu — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: all files reviewed, 2 unresolved discussions (waiting on @wangdayong228)

**Review by boqiu — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-:-MaYHwbm3_-yASJTHHwE:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @wangdayong228)

**Review by wangdayong228 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @wangdayong228)

___
*[middleware/call_rpc.go, line 35 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/78#-MaWMIz852eWynBvrdxN:-MaYHxMV82a9M60mq6w6:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/3b60cab5d03b2b2f22ab2c84e7eb224e50cdc324/middleware/call_rpc.go#L35)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Generally, the default behavior should not print any error message in stdout. Instead, you could return a stacked error or add any debug level log.

Just remind, no app will print error message to console by default.
</blockquote></details>

Done.
